src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+ */
+static void
+RenameRelationToShardRelation(Oid shellRelationId, uint64 shardId)
+{
+	Oid shellRelationSchemaId = get_rel_namespace(shellRelationId);
+	char *shellRelationSchemaName = get_namespace_name(shellRelationSchemaId);
+	char *shellRelationName = get_rel_name(shellRelationId);
+	char *qualifiedShellRelationName = quote_qualified_identifier(shellRelationSchemaName,
+																  shellRelationName);
+
+	char *shardRelationName = pstrdup(shellRelationName);
+	AppendShardIdToName(&shardRelationName, shardId);
+
+	StringInfo renameCommand = makeStringInfo();
+	appendStringInfo(renameCommand, "ALTER TABLE %s RENAME TO %s;",
+					 qualifiedShellRelationName, shardRelationName);
+
+	const char *commandString = renameCommand->data;
+
+	Node *parseTree = ParseTreeNode(commandString);
+
+	CitusProcessUtility(parseTree, commandString, PROCESS_UTILITY_TOPLEVEL,
+						NULL, None_Receiver, NULL);
+}
+
+
+/*
+ * RenameShardRelationConstraints appends given shardId to the end of the name
+ * of constraints "defined on" the relation with shardRelationId. This function
+ * utilizes GetConstraintNameList to pick the constraints to be renamed,
+ * see more details in function's comment.
+ */
+static void
+RenameShardRelationConstraints(Oid shardRelationId, uint64 shardId)
+{
+	Oid shardRelationSchemaId = get_rel_namespace(shardRelationId);
+	char *shardRelationSchemaName = get_namespace_name(shardRelationSchemaId);
+	char *shardRelationName = get_rel_name(shardRelationId);
+	char *qualifiedShardRelationName =
+		quote_qualified_identifier(shardRelationSchemaName, shardRelationName);
+
+	List *constraintNameList = GetConstraintNameList(shardRelationId);
+
+	char *constraintName = NULL;
+	foreach_ptr(constraintName, constraintNameList)
+	{
+		char *shardConstraintName = pstrdup(constraintName);
+		AppendShardIdToName(&shardConstraintName, shardId);
+
+		StringInfo renameCommand = makeStringInfo();
+		appendStringInfo(renameCommand, "ALTER TABLE %s RENAME CONSTRAINT %s TO %s;",
+						 qualifiedShardRelationName, constraintName, shardConstraintName);
+
+		const char *commandString = renameCommand->data;
+
+		Node *parseTree = ParseTreeNode(commandString);
+
+		CitusProcessUtility(parseTree, commandString, PROCESS_UTILITY_TOPLEVEL,
+							NULL, None_Receiver, NULL);
+	}
+}
+
+
+/*
+ * GetConstraintNameList returns a list constraint names "defined on" the
+ * relation with relationId. Those constraints can be:
+ *  - "check" constraints or,
+ *  - "primary key" constraints or,
+ *  - "unique" constraints or,
+ *  - "trigger" constraints or,
+ *  - "exclusion" constraints or,
+ *  - "foreign key" constraints in which the relation is the "referencing"
+ *     relation (including the self-referencing foreign keys).
+ */
+static List *
+GetConstraintNameList(Oid relationId)
+{
+	List *constraintNameList = NIL;
+
+	int scanKeyCount = 1;
+	ScanKeyData scanKey[1];
+
+	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
```

Comment:
Actually this function does also scan `pg_constraint` but it does that for all kind of constrains on the contrary of the other two functions related in `foreign_constraints.c`. So I mean, in terms of being a function related with all types constraints, this is the only function.

It might still make sense to implement an internal method first to return all kind of constraints and then to use that method in `foreign_constraints.c` too, but I'm not sure if that refactor should be done in a seperate pr first ?

Reply:
Yeah I think it makes sense to do it in another PR, so the current code seems fine with the current version of that function